### DISCUSSION TITLE
make config tolerant to no value

### DISCUSSION THREAD

**Discussion by rafikt1992** [k14s/ytt#445] [Q&A] [answered]
Hello,

I have been searching for a while to solve this problem

if my config.yaml is 

```
#@ load("@ytt:data", "data")
foo: data.value.foo
bar: data.value.bar
```

and my value file is 
```
#@data/values
---
foo: baz
```

how can I make my config.yam tolerate not defining `bar` in my value file without throwing an error `structure has no xx field or method.` and simply discard the key in the output file.

thank you

Edit:

I found a solution finally. One way to do it is to make a map of all the available keys in the value file since the value file is a struct using struct function ` mymap= dir(data.values)` then test if the key exist. `config.yaml` should look like this:

```
#@ load("@ytt:data", "data")
 mymap= dir(data.values)
foo: data.value.foo
#@ if/end 'bar' in mymap:
bar: data.value.bar

```

**Comment by gcheadle-vmware** [comment, marked as answer]
Hi @rafikt1992 👋 

I see your edit and that you have found a solution 👍  
Some helpful documentation for this question lives in our [FAQ doc under the Data Values section](https://carvel.dev/ytt/docs/latest/faq/#data-values).
One note is that `@ data.values` is already a struct, thus it has a function to test the existence of a key, this function is [`hasattr()`](https://github.com/google/starlark-go/blob/master/doc/spec.md#hasattr):
```
#@ load("@ytt:data", "data")
foo: #@ data.value.foo
#@ if/end hasattr(data.values, "bar"):
bar: #@ data.value.bar
```
